assert/assertions.go

```diff
@@ -54,116 +55,28 @@ type Comparison func() (success bool)
 
 // ObjectsAreEqual determines if two objects are considered equal.
 //
-// This function does no assertion of any kind.
+// Deprecated: This function does no assertion of any kind. Use Equal instead.
 func ObjectsAreEqual(expected, actual interface{}) bool {
-	if expected == nil || actual == nil {
-		return expected == actual
-	}
-
-	exp, ok := expected.([]byte)
-	if !ok {
-		return reflect.DeepEqual(expected, actual)
-	}
-
-	act, ok := actual.([]byte)
-	if !ok {
-		return false
-	}
-	if exp == nil || act == nil {
-		return exp == nil && act == nil
-	}
-	return bytes.Equal(exp, act)
-}
-
-// copyExportedFields iterates downward through nested data structures and creates a copy
-// that only contains the exported struct fields.
-func copyExportedFields(expected interface{}) interface{} {
-	if isNil(expected) {
-		return expected
-	}
-
-	expectedType := reflect.TypeOf(expected)
-	expectedKind := expectedType.Kind()
-	expectedValue := reflect.ValueOf(expected)
-
-	switch expectedKind {
-	case reflect.Struct:
-		result := reflect.New(expectedType).Elem()
-		for i := 0; i < expectedType.NumField(); i++ {
-			field := expectedType.Field(i)
-			isExported := field.IsExported()
-			if isExported {
-				fieldValue := expectedValue.Field(i)
-				if isNil(fieldValue) || isNil(fieldValue.Interface()) {
-					continue
-				}
-				newValue := copyExportedFields(fieldValue.Interface())
-				result.Field(i).Set(reflect.ValueOf(newValue))
-			}
-		}
-		return result.Interface()
-
-	case reflect.Ptr:
-		result := reflect.New(expectedType.Elem())
-		unexportedRemoved := copyExportedFields(expectedValue.Elem().Interface())
-		result.Elem().Set(reflect.ValueOf(unexportedRemoved))
-		return result.Interface()
-
-	case reflect.Array, reflect.Slice:
-		result := reflect.MakeSlice(expectedType, expectedValue.Len(), expectedValue.Len())
-		for i := 0; i < expectedValue.Len(); i++ {
-			index := expectedValue.Index(i)
-			if isNil(index) {
-				continue
-			}
-			unexportedRemoved := copyExportedFields(index.Interface())
-			result.Index(i).Set(reflect.ValueOf(unexportedRemoved))
-		}
-		return result.Interface()
-
-	case reflect.Map:
-		result := reflect.MakeMap(expectedType)
-		for _, k := range expectedValue.MapKeys() {
-			index := expectedValue.MapIndex(k)
-			unexportedRemoved := copyExportedFields(index.Interface())
-			result.SetMapIndex(k, reflect.ValueOf(unexportedRemoved))
-		}
-		return result.Interface()
-
-	default:
-		return expected
-	}
+	return check.ObjectsAreEqual(expected, actual)
 }
 
 // ObjectsExportedFieldsAreEqual determines if the exported (public) fields of two objects are
 // considered equal. This comparison of only exported fields is applied recursively to nested data
 // structures.
 //
-// This function does no assertion of any kind.
+// Deprecated: This function does no assertion of any kind. Use
+// EqualExportedValues instead.
 func ObjectsExportedFieldsAreEqual(expected, actual interface{}) bool {
```

Comment:
The deprecation message must suggest a full alternative to ease migrating away.

Reply:
The signature shares the same arguments, no?

Reply:
In #1279 you hinted that people should use go-cmp instead. However the testify functions have a particular definition of equality as `nil` values in the *expected* value mean *ignore this field*.

I think that we have the responsibility of helping users to find that path if we are deprecating the functions.

So, do we have a path forward for all use cases? I think we must provide that detailed information in the documentation. If not, we will be get loads of support questions.

Reply:
Making a specific recommendation, such as go-cmp, is something we'd have to maintain though. What if go-cmp declares itself EOL tomorrow? I'd be happy to clarify that one should use EqualExportedValues if you intend to make an assertion and you should use another module if you intend to make a comparison. But I'm not happy to make any specific recommendation.

Reply:
If we have no safe migration to suggest we must not deprecate.

Reply:
I think removing functions which make no assertion from Assert, so that people using testify to test don't have hidden bugs is more important than providing a forward path for people using testify as a utility library. The 4th paragraph of [Philosophy](https://github.com/stretchr/testify/blob/master/CONTRIBUTING.md#philosophy) supports this.

Reply:
Some development teams have policies which disallow using deprecated functions/methods. If we don't provide a path forward some users might be tempted to stop upgrading Testify to avoid the deprecation warning instead of dropping the use of `ObjectsAreEquals`.

I don't think that this change is worth a fight with our users.